### PR TITLE
rubocops/lines: allow a runtime Python when supporting multiple Pythons

### PR DESCRIPTION
**Pull request by cho-m**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Draft for now as unable to test out on Sonoma and not too sure about RuboCop logic. Maybe there is better way to handle.

Main thing is to handle recent changes to various formulae where multiple Python libraries were built and runtime dependency was dropped resulting in non-functioning executables installed/linked into user PATH, e.g. `pyinvoke` (https://github.com/Homebrew/homebrew-core/issues/151248), `python-tabulate`, `python-argcomplete`, etc...

### CODE REVIEW

File: Library/Homebrew/rubocops/lines.rb [hunk 1]
```diff
@@ -379,11 +379,21 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
...
+
+            dep
+          end.compact
+
           find_strings(body_node).each do |str|
             content = string_content(str)
 
             next unless (match = content.match(/^python(@)?(\d\.\d+)$/))
             next if python_version == match[2]
+            next if other_python_deps.include?(str)
```
Comment (cho-m) [inline Library/Homebrew/rubocops/lines.rb:396]: Will either need to check way to check same object as this just checks same string right now. Other way I was originally considering was checking if `str` is part of `depends_on` but wasn't sure if easy way to get this.

**Comment by Bo98** [comment]
This makes sense in isolation.

Though it does seem problematic when we have any dependents using it as we'll get mixed runtime Pythons again and they will now need to be migrated as one.

Alternative solutions could be:
* Have some shim that automatically picks whatever supported Python it can find in the PATH.
* Separate executables into a separate formula (perhaps tricky given we don't support subpackages)

**Comment by cho-m** [comment]
> Though it does seem problematic when we have any dependents using it as we'll get mixed runtime Pythons again and they will now need to be migrated as one.

As I recall, we ignore mixed Pythons in dependency tree (#13714) so migration should still be possible separately.

> Alternative solutions could be:
> 
> * Introduce a shim executable that automatically picks whatever supported Python it can find in the PATH.

I've thought on this before but seemed like it would end up a bit hacky. Open to discussion if there is a reasonable way. However, some formulae will still be non-functional out-of-the-box unless user installs a python separately (which may require propagating caveats to impacted formulae).

> * Separate executables into a separate formula (perhaps tricky given we don't support subpackages)

We've tried this before and reverted those formulae now (e.g. `python-tabulate` + `libpython-tabulate`). It often required various workarounds to support.

**Comment by Bo98** [comment]
> As I recall, we ignore mixed Pythons in dependency tree (#13714) so migration should still be possible separately.

It's not a great user experience though for a formula to depend on two versions of Python.

**Comment by cho-m** [comment]
> It's not a great user experience though for a formula to depend on two versions of Python, particularly when only one is really needed.

Ideally, should only happen for short time during migration.

We probably should discuss/document further on Python package distribution as we don't seem to be following our original guidelines, e.g. https://docs.brew.sh/Python-for-Formula-Authors

---

> I'm fine with this change if it's preferred, but it does feel like there's better solutions.

I am open to any other ideas. Just that during current `python@3.11 --> python@3.12` migration we have impacted more formulae than previous migrations so may end up seeing more user issues.

---

An opposite approach would be preventing "multiple pythons" for formulae whose functionality depends on the bin scripts.

For example,
- `pyinvoke` workflow needs bin scripts - https://docs.pyinvoke.org/en/stable/getting-started.html#defining-and-running-task-functions (https://github.com/Homebrew/homebrew-core/issues/151248) 
- I think `docutils` also commonly uses bin scripts - https://docutils.sourceforge.io/README.html#usage

On other hand, there are some formulae that have bin scripts that are not commonly used (e.g. `python-tabulate`).

**Comment by MikeMcQuaid** [comment]
> We probably should discuss/document further on Python package distribution as we don't seem to be following our original guidelines, e.g. https://docs.brew.sh/Python-for-Formula-Authors

If we can turn any of those guidelines into code/audits/rubocops: that would be ideal.



> An opposite approach would be preventing "multiple pythons" for formulae whose functionality depends on the bin scripts.

This makes sense to me 👍🏻

**Comment by cho-m** [comment]
Will close for now until we finalize on how we want to handle Python formulae moving forward (especially given adoption of PEP 668 in `python@3.12` and other changes ongoing in `homebrew-core`).

**Pull request closed without merging**